### DISCUSSION TITLE
Spacy 3.o DocBin.to_disk

### DISCUSSION THREAD

**Discussion by srikar-kodati** [explosion/spaCy#7979] [Help: Coding & Implementations] [answered]
Hello all!

I have this bit of code:

```
nlp = spacy.blank("en")
def create_training(TRAIN_DATA):
    db = DocBin()
    for text, annot in tqdm(TRAIN_DATA):
        doc = nlp.make_doc(text)
        ents = []
        for start,end,label in annot['entities']:
            span = doc.char_span(start,end, label=label, alignment_mode = 'contract')
            if span is None:
                print("Skipping entity")
            else:
                ents.append(span)
        doc.ents = ents
        db.add(doc)
    return (db)


train = create_training(TRAIN_DATA)
train.to_disk("./data/train.spacy")
```

But i seem to get this error message:
'DocBin' object has no attribute 'to_disk'

Why isn't that working?? I cannot find an alternative to this

**Comment by polm** [comment]
That's weird, your code works for me if I remove the for loop and add imports:

```
import spacy
from spacy.tokens import DocBin
```

Can you share your complete code and the output of `pip freeze`?

**Comment by srikar-kodati** [comment, marked as answer]
pip freeze
aiohttp @ file:///D:/bld/aiohttp_1614281086591/work
alabaster==0.7.12
anaconda-client==1.7.2
anaconda-navigator==2.0.1
anaconda-project==0.8.3
argh==0.26.2
argon2-cffi @ file:///C:/ci/argon2-cffi_1596828585465/work
asn1crypto @ file:///tmp/build/80754af9/asn1crypto_1596577642040/work
astroid @ file:///C:/ci/astroid_1592487315634/work
astropy==4.0.2
async-generator==1.10
async-timeout==3.0.1
atomicwrites==1.4.0
attrs @ file:///tmp/build/80754af9/attrs_1604765588209/work
autopep8 @ file:///tmp/build/80754af9/autopep8_1596578164842/work
Babel @ file:///tmp/build/80754af9/babel_1605108370292/work
backcall==0.2.0
backports.functools-lru-cache==1.6.1
backports.shutil-get-terminal-size==1.0.0
backports.tempfile==1.0
backports.weakref==1.0.post1
bcrypt @ file:///C:/ci/bcrypt_1597936263757/work
beautifulsoup4 @ file:///tmp/build/80754af9/beautifulsoup4_1601924105527/work
bitarray @ file:///C:/ci/bitarray_1605065210072/work
bkcharts==0.2
bleach @ file:///tmp/build/80754af9/bleach_1600439572647/work
blis @ file:///D:/bld/cython-blis_1607338345852/work
bokeh @ file:///C:/ci/bokeh_1603297934731/work
boto==2.49.0
boto3 @ file:///home/conda/feedstock_root/build_artifacts/boto3_1619570519752/work
botocore @ file:///home/conda/feedstock_root/build_artifacts/botocore_1619588579232/work
Bottleneck==1.3.2
brotlipy==0.7.0
bz2file==0.98
cachetools @ file:///home/conda/feedstock_root/build_artifacts/cachetools_1619593780156/work
catalogue @ file:///D:/bld/catalogue_1618488991174/work
certifi==2020.6.20
cffi @ file:///C:/ci/cffi_1600699246375/work
chardet==3.0.4
click==7.1.2
cloudpickle @ file:///tmp/build/80754af9/cloudpickle_1598884132938/work
clyent==1.2.2
colorama @ file:///tmp/build/80754af9/colorama_1603211150991/work
comtypes==1.1.7
conda==4.10.1
conda-build==3.20.5
conda-package-handling @ file:///C:/ci/conda-package-handling_1603003327818/work
conda-repo-cli @ file:///tmp/build/80754af9/conda-repo-cli_1611347565866/work
conda-token @ file:///tmp/build/80754af9/conda-token_1613597706833/work
conda-verify==3.4.2
contextlib2==0.6.0.post1
cryptography @ file:///C:/ci/cryptography_1601046905460/work
cycler==0.10.0
cymem @ file:///D:/bld/cymem_1607336390769/work
Cython @ file:///C:/ci/cython_1594829190914/work
cytoolz==0.11.0
dask @ file:///tmp/build/80754af9/dask-core_1602083700509/work
decorator==4.4.2
defusedxml==0.6.0
diff-match-patch @ file:///tmp/build/80754af9/diff-match-patch_1594828741838/work
distributed @ file:///C:/ci/distributed_1605066635297/work
docutils==0.16
en-core-web-lg @ https://github.com/explosion/spacy-models/releases/download/en_core_web_lg-3.0.0/en_core_web_lg-3.0.0-py3-none-any.whl
en-core-web-md @ https://github.com/explosion/spacy-models/releases/download/en_core_web_md-3.0.0/en_core_web_md-3.0.0-py3-none-any.whl
en-core-web-sm @ file:///home/conda/feedstock_root/build_artifacts/spacy-model-en_core_web_1612236387798/work/sm
entrypoints==0.3
et-xmlfile==1.0.1
fastcache==1.1.0
filelock==3.0.12
flake8 @ file:///tmp/build/80754af9/flake8_1601911421857/work
Flask==1.1.2
fsspec @ file:///tmp/build/80754af9/fsspec_1602684995936/work
future==0.18.2
gevent @ file:///C:/ci/gevent_1601397627942/work
glob2==0.7
google-api-core @ file:///home/conda/feedstock_root/build_artifacts/google-api-core-split_1616709747968/work
google-auth @ file:///home/conda/feedstock_root/build_artifacts/google-auth_1616708961589/work
google-cloud-core @ file:///home/conda/feedstock_root/build_artifacts/google-cloud-core_1608136972897/work
google-cloud-storage @ file:///home/conda/feedstock_root/build_artifacts/google-cloud-storage_1616708661781/work
google-crc32c @ file:///D:/bld/google-crc32c_1611954423366/work
google-resumable-media @ file:///home/conda/feedstock_root/build_artifacts/google-resumable-media_1608141079349/work
googleapis-common-protos @ file:///D:/bld/googleapis-common-protos-feedstock_1616711070963/work
greenlet @ file:///C:/ci/greenlet_1600885448389/work
grpcio @ file:///D:/bld/grpcio_1607005218920/work
h5py==2.10.0
HeapDict==1.0.1
html5lib @ file:///tmp/build/80754af9/html5lib_1593446221756/work
idna @ file:///tmp/build/80754af9/idna_1593446292537/work
imageio @ file:///tmp/build/80754af9/imageio_1594161405741/work
imagesize==1.2.0
importlib-metadata @ file:///tmp/build/80754af9/importlib-metadata_1602276842396/work
iniconfig @ file:///tmp/build/80754af9/iniconfig_1602780191262/work
intervaltree @ file:///tmp/build/80754af9/intervaltree_1598376443606/work
ipykernel @ file:///C:/ci/ipykernel_1596190155316/work/dist/ipykernel-5.3.4-py3-none-any.whl
ipython @ file:///C:/ci/ipython_1604083276484/work
ipython-genutils==0.2.0
ipywidgets @ file:///tmp/build/80754af9/ipywidgets_1601490159889/work
isort @ file:///tmp/build/80754af9/isort_1602603989581/work
itsdangerous==1.1.0
jdcal==1.4.1
jedi @ file:///C:/ci/jedi_1592833825077/work
Jinja2==2.11.2
jmespath @ file:///home/conda/feedstock_root/build_artifacts/jmespath_1589369830981/work
joblib @ file:///tmp/build/80754af9/joblib_1601912903842/work
json5==0.9.5
jsonschema @ file:///tmp/build/80754af9/jsonschema_1602607155483/work
jupyter==1.0.0
jupyter-client @ file:///tmp/build/80754af9/jupyter_client_1601311786391/work
jupyter-console @ file:///tmp/build/80754af9/jupyter_console_1598884538475/work
jupyter-core==4.6.3
jupyterlab==2.2.6
jupyterlab-pygments @ file:///tmp/build/80754af9/jupyterlab_pygments_1601490720602/work
jupyterlab-server @ file:///tmp/build/80754af9/jupyterlab_server_1594164409481/work
keyring @ file:///C:/ci/keyring_1601489747083/work
kiwisolver @ file:///C:/ci/kiwisolver_1604014703538/work
lazy-object-proxy==1.4.3
libarchive-c==2.9
llvmlite==0.34.0
locket==0.2.0
lxml @ file:///C:/ci/lxml_1603216366346/work
MarkupSafe==1.1.1
matplotlib @ file:///C:/ci/matplotlib-base_1603355780617/work
mccabe==0.6.1
menuinst==1.4.16
mistune==0.8.4
mkl-fft==1.2.0
mkl-random==1.1.1
mkl-service==2.3.0
mock==4.0.2
more-itertools @ file:///tmp/build/80754af9/more-itertools_1605111547926/work
mpmath==1.1.0
msgpack==1.0.0
multidict @ file:///D:/bld/multidict_1610319170773/work
multipledispatch==0.6.0
murmurhash @ file:///D:/bld/murmurhash_1607334400096/work
navigator-updater==0.2.1
nbclient @ file:///tmp/build/80754af9/nbclient_1602783176460/work
nbconvert @ file:///C:/ci/nbconvert_1601914925608/work
nbformat @ file:///tmp/build/80754af9/nbformat_1602783287752/work
nest-asyncio @ file:///tmp/build/80754af9/nest-asyncio_1605115881283/work
networkx @ file:///tmp/build/80754af9/networkx_1598376031484/work
nltk @ file:///tmp/build/80754af9/nltk_1592496090529/work
nose==1.3.7
notebook @ file:///C:/ci/notebook_1601501643625/work
numba==0.51.2
numexpr==2.7.1
numpy @ file:///C:/ci/numpy_and_numpy_base_1603466732592/work
numpydoc @ file:///tmp/build/80754af9/numpydoc_1605117425582/work
olefile==0.46
openpyxl @ file:///tmp/build/80754af9/openpyxl_1598113097404/work
packaging==20.4
pandas @ file:///C:/ci/pandas_1602083338010/work
pandocfilters @ file:///C:/ci/pandocfilters_1605102497129/work
paramiko @ file:///tmp/build/80754af9/paramiko_1598886428689/work
parso==0.7.0
partd==1.1.0
path @ file:///C:/ci/path_1598358556930/work
pathlib2==2.3.5
pathtools==0.1.2
pathy @ file:///home/conda/feedstock_root/build_artifacts/pathy_1619290361841/work
patsy==0.5.1
pep8==1.7.1
pexpect==4.8.0
pickleshare==0.7.5
Pillow @ file:///C:/ci/pillow_1603823068645/work
pkginfo==1.6.1
pluggy==0.13.1
ply==3.11
preshed @ file:///D:/bld/preshed_1607337932670/work
prometheus-client==0.8.0
prompt-toolkit @ file:///tmp/build/80754af9/prompt-toolkit_1602688806899/work
protobuf==3.15.8
psutil @ file:///C:/ci/psutil_1598370330503/work
py @ file:///tmp/build/80754af9/py_1593446248552/work
pyasn1==0.4.8
pyasn1-modules==0.2.7
pycodestyle==2.6.0
pycosat==0.6.3
pycparser @ file:///tmp/build/80754af9/pycparser_1594388511720/work
pycurl==7.43.0.6
pydantic @ file:///D:/bld/pydantic_1614365346102/work
pydocstyle @ file:///tmp/build/80754af9/pydocstyle_1598885001695/work
pyflakes==2.2.0
Pygments @ file:///tmp/build/80754af9/pygments_1604103097372/work
pylint @ file:///C:/ci/pylint_1598617153160/work
PyNaCl @ file:///C:/ci/pynacl_1595000047588/work
pyodbc===4.0.0-unsupported
pyOpenSSL @ file:///tmp/build/80754af9/pyopenssl_1594392929924/work
pyparsing==2.4.7
pyreadline==2.1
pyrsistent @ file:///C:/ci/pyrsistent_1600141795814/work
PySocks==1.7.1
pytest==0.0.0
python-dateutil==2.8.1
python-jsonrpc-server @ file:///tmp/build/80754af9/python-jsonrpc-server_1600278539111/work
python-language-server @ file:///tmp/build/80754af9/python-language-server_1600454544709/work
pytz==2020.1
PyWavelets @ file:///C:/ci/pywavelets_1601658407916/work
pywin32==227
pywin32-ctypes==0.2.0
pywinpty==0.5.7
PyYAML==5.3.1
pyzmq==19.0.2
QDarkStyle==2.8.1
QtAwesome @ file:///tmp/build/80754af9/qtawesome_1602272867890/work
qtconsole @ file:///tmp/build/80754af9/qtconsole_1600870028330/work
QtPy==1.9.0
regex @ file:///C:/ci/regex_1602770567298/work
requests @ file:///tmp/build/80754af9/requests_1592841827918/work
rope @ file:///tmp/build/80754af9/rope_1602264064449/work
rsa @ file:///home/conda/feedstock_root/build_artifacts/rsa_1614171254180/work
Rtree==0.9.4
ruamel-yaml==0.15.87
s3transfer @ file:///home/conda/feedstock_root/build_artifacts/s3transfer_1619138243643/work
scikit-image==0.17.2
scikit-learn @ file:///C:/ci/scikit-learn_1598377018496/work
scipy @ file:///C:/ci/scipy_1604596260408/work
seaborn @ file:///tmp/build/80754af9/seaborn_1600553570093/work
Send2Trash==1.5.0
shellingham @ file:///home/conda/feedstock_root/build_artifacts/shellingham_1612179560728/work
simplegeneric==0.8.1
singledispatch @ file:///tmp/build/80754af9/singledispatch_1602523705405/work
sip==4.19.13
six @ file:///C:/ci/six_1605187374963/work
smart-open==3.0.0
snowballstemmer==2.0.0
sortedcollections==1.2.1
sortedcontainers==2.2.2
soupsieve==2.0.1
spacey==0.1.1
spacy @ file:///D:/bld/spacy_1619254547217/work
spacy-legacy @ file:///home/conda/feedstock_root/build_artifacts/spacy-legacy_1619277335881/work
Sphinx @ file:///tmp/build/80754af9/sphinx_1597428793432/work
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==1.0.3
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.4
sphinxcontrib-websupport @ file:///tmp/build/80754af9/sphinxcontrib-websupport_1597081412696/work
spyder @ file:///C:/ci/spyder_1599054115447/work
spyder-kernels @ file:///C:/ci/spyder-kernels_1599051709837/work
SQLAlchemy @ file:///C:/ci/sqlalchemy_1603818908746/work
srsly @ file:///D:/bld/srsly_1618231808282/work
statsmodels==0.12.0
sympy @ file:///C:/ci/sympy_1605101511098/work
tables==3.6.1
tblib @ file:///tmp/build/80754af9/tblib_1597928476713/work
terminado==0.9.1
testpath==0.4.4
thinc @ file:///D:/bld/thinc_1619212195627/work
threadpoolctl @ file:///tmp/tmp9twdgx9k/threadpoolctl-2.1.0-py3-none-any.whl
tifffile==2020.10.1
toml @ file:///tmp/build/80754af9/toml_1592853716807/work
toolz @ file:///tmp/build/80754af9/toolz_1601054250827/work
tornado==6.0.4
tqdm @ file:///tmp/build/80754af9/tqdm_1602185206534/work
traitlets @ file:///tmp/build/80754af9/traitlets_1602787416690/work
typer @ file:///home/conda/feedstock_root/build_artifacts/typer_1609874382867/work
typing-extensions @ file:///tmp/build/80754af9/typing_extensions_1598376058250/work
ujson @ file:///C:/ci/ujson_1602505266082/work
unicodecsv==0.14.1
urllib3 @ file:///tmp/build/80754af9/urllib3_1603305693037/work
wasabi @ file:///home/conda/feedstock_root/build_artifacts/wasabi_1612156086016/work
watchdog @ file:///C:/ci/watchdog_1593447437088/work
wcwidth @ file:///tmp/build/80754af9/wcwidth_1593447189090/work
webencodings==0.5.1
Werkzeug==1.0.1
widgetsnbextension==3.5.1
win-inet-pton==1.1.0
win-unicode-console==0.5
wincertstore==0.2
wordcloud==1.8.1
wrapt==1.11.2
xlrd==1.2.0
XlsxWriter @ file:///tmp/build/80754af9/xlsxwriter_1602692860603/work
xlwings==0.20.8
xlwt==1.3.0
xmltodict==0.12.0
yapf @ file:///tmp/build/80754af9/yapf_1593528177422/work
yarl @ file:///D:/bld/yarl_1603108567357/work
zict==2.0.0
zipp @ file:///tmp/build/80754af9/zipp_1604001098328/work
zope.event==4.5.0
zope.interface @ file:///C:/ci/zope.interface_1602002494740/work










## OLD SPACY FORMAT
## In order to work on spacy 3.o... this has to be converted to .spacy by converting into doc and then a docbin

TRAIN_DATA = [('The F15 aircraft uses a lot of fuel', {'entities': [(4, 7, 'aircraft')]}),
 ('did you see the F16 landing?', {'entities': [(16, 19, 'aircraft')]}),
 ('how many missiles can a F35 carry', {'entities': [(24, 27, 'aircraft')]}),
 ('is the F15 outdated', {'entities': [(7, 10, 'aircraft')]}),
 ('does the US still train pilots to dog fight?',
  {'entities': [(0, 0, 'aircraft')]}),
 ('how long does it take to train a F16 pilot',
  {'entities': [(33, 36, 'aircraft')]}),
 ('how much does a F35 cost', {'entities': [(16, 19, 'aircraft')]}),
 ('would it be possible to steal a F15', {'entities': [(32, 35, 'aircraft')]}),
 ('who manufactures the F16', {'entities': [(21, 24, 'aircraft')]}),
 ('how many countries have bought the F35',
  {'entities': [(35, 38, 'aircraft')]}),
 ('is the F35 a waste of money', {'entities': [(7, 10, 'aircraft')]})]




import pandas as pd
from tqdm import tqdm
import spacy
from spacy.tokens import DocBin

nlp = spacy.blank("en") # load a new spacy model
db = DocBin() # create a DocBin object

for text, annot in tqdm(TRAIN_DATA): # data in previous format
    doc = nlp.make_doc(text) # create doc object from text
    ents = []
    for start, end, label in annot["entities"]: # add character indexes
        span = doc.char_span(start, end, label=label, alignment_mode="contract")
        if span is None:
            print("Skipping entity")
        else:
            ents.append(span)
    doc.ents = ents # label the text with the ents
    db.add(doc)

db.to_disk("./train.spacy") # save the docbin object

**Reply by adrianeboyd** [reply]
Hi, it looks like you have a lot of packages that are compiled from source. For `spacy`, this might be an unstable or alpha version built from the `develop` branch before any of the v3.0.0rc nightly releases?

`DocBin.to_disk` was added in #5892, which as far as I remember was well before the version v3.0.0rc0 release. Please try again with the most recent release v3.0.6 or see if you can update your source install?

**Reply by srikar-kodati** [reply]
To clarify, youre saying to update spacy?

**Reply by adrianeboyd** [reply]
Yes, that's probably the solution. From this `pip freeze` output we can't tell which version you have installed, so either something is wrong with the installation in general (although that's not that likely given this behavior), or somehow you have a pre-release version of v3.0. `spacy.git_info.GIT_VERSION` could potentially help you figure out how old it is.

**Reply by srikar-kodati** [reply]
I have spacy version 3.0.6

**Reply by adrianeboyd** [reply]
How did you install spacy?

For comparison, can you create a new virtual environment and install spacy from `conda-forge` instead to test?

```bash
conda install spacy -c conda-forge
```

**Reply by srikar-kodati** [reply]
I opened my anaconda command prompt and followed the instruction given by spacy:

pip install -U spacy
python -m spacy download en_core_web_sm
python -m spacy download en_core_web_md
python -m spacy download en_core_web_lg


I couldnt get the virtual environment way to work. Idk why, it doesnt work for me

**Reply by adrianeboyd** [reply]
Hmm, I'm sorry I can't really tell what's going on from the output above. We can try to provide support for the installation methods here: https://spacy.io/usage

If you're using anaconda, try creating a new environment with `conda create` and then installing from `conda-forge` as shown in the quickstart.

I see that the quick start "virtual env" instructions accidentally mix up pip and conda a bit. We can get that updated soon! Use `conda create`, `conda activate` and then the simple `conda install` instructions.

**Reply by srikar-kodati** [reply]
May be an odd request. Am I able to screen share with you via teams or something? I cant seem to fix the issue

**Reply by adrianeboyd** [reply]
Sorry, we're not able to do that. I think you'll need to figure out how to get a new clean virtual environment, either with pip or conda, and after you've activated it, start from there with our install instructions.

**Reply by srikar-kodati** [reply]
I reinstalled everything and got it to work. Thank you so much! Spacy is such a powerful tool

**Reply by adrianeboyd** [reply]
Glad to hear it!

**Comment by srikar-kodati** [comment]
To clarify, youre saying to update spacy?
